### ISSUE TITLE
ControlFlow::Wait suspends event if redraw is requested.

### ISSUE THREAD

**Issue by zakarumych** [rust-windowing/winit#1619]
I set flow to `ControlFlow::Wait` in order to not waste CPU on doing nothing.
When application calls `Window::request_redraw()` I would expect `RedrawRequest` to arrive right away (after `MainRequestCleared` if there are queued events). Yet no `ReadrawRequest` is received until other events happen. Particularly if mouse and keyboard are untouched app would wait for `ReadrawRequest` indefinitely.

I tried to call `Window::request_redraw()` at different moments. Nothing helps. Then I found that using `ControlFlow::Poll` causes `RedrawRequest` to be received as expected, without any input.

Is it expected behavior? I couldn't find anything like this in the docs for `ControlFlow`.

Platform: Windows 10.

**Comment by Osspial** [comment]
Could you provide some example code and relevant console logs? I’m unable to reproduce the issue you’re describing with this code:

```rust
    event::{ElementState, Event, WindowEvent},
    event_loop::{ControlFlow, EventLoop},
    window::WindowBuilder,
};

fn main() {
    simple_logger::init().unwrap();
    let event_loop = EventLoop::new();

    let window = WindowBuilder::new()
        .with_title("A fantastic window!")
        .build(&event_loop)
        .unwrap();

    event_loop.run(move |event, _, control_flow| {
        println!("{:?}", event);

        *control_flow = ControlFlow::Wait;

        match event {
            Event::WindowEvent { event, .. } => match event {
                WindowEvent::CloseRequested => *control_flow = ControlFlow::Exit,
                _ => (),
            },
            Event::MainEventsCleared => {
                window.request_redraw();
            }
            Event::RedrawRequested(_) => {
                println!("\nredrawing!\n");
            }
            _ => (),
        }
    });
}
```

**Comment by zakarumych** [comment]
I copied your snippet into crate (with little modifications) and it waits for events.
https://github.com/zakarumych/winit-wait-bug

What additional info I can provide to you?

**Comment by zakarumych** [comment]
A bit of debugging revealed this.

After some events `thread_event_target_callback` handles `WM_PAINT`, which is followed by `flush_paint_messages` to peek WM_PAINT message and dispatch it, causing `public_window_callback` to receive `WM_PAINT` and generate `RedrawRequested` event.

After that, if no input occur, `windows::event_loop::thread_event_target_callback` receives `WM_PAINT` and that's it.

Logs look like this: 
```
2020-07-16 14:03:05,781 INFO  [winit_wait] DeviceEvent { device_id: DeviceId(DeviceId(65596)), event: Motion { axis: 0, value: -1.0 } }
2020-07-16 14:03:05,782 INFO  [winit_wait] DeviceEvent { device_id: DeviceId(DeviceId(65596)), event: MouseMotion { delta: (-1.0, 0.0) } }
2020-07-16 14:03:05,782 INFO  [winit_wait] DeviceEvent { device_id: DeviceId(DeviceId(65596)), event: Motion { axis: 0, value: -1.0 } }
2020-07-16 14:03:05,782 INFO  [winit_wait] DeviceEvent { device_id: DeviceId(DeviceId(65596)), event: MouseMotion { delta: (-1.0, 0.0) } }
2020-07-16 14:03:05,782 ERROR [winit::platform_impl::platform::event_loop] Thread WM_PAINT received
2020-07-16 14:03:05,783 INFO  [winit_wait] MainEventsCleared
2020-07-16 14:03:05,783 ERROR [winit::platform_impl::platform::event_loop] Peeked WM_PAINT message
2020-07-16 14:03:05,783 ERROR [winit::platform_impl::platform::event_loop] Window WM_PAINT received
2020-07-16 14:03:05,783 INFO  [winit_wait] RedrawRequested(WindowId(WindowId(0x280c68)))
2020-07-16 14:03:05,783 INFO  [winit_wait] redrawing!
2020-07-16 14:03:05,783 INFO  [winit_wait] RedrawEventsCleared
2020-07-16 14:03:05,784 ERROR [winit::platform_impl::platform::event_loop] Thread WM_PAINT received
```

**Comment by parasyte** [comment]
Seeing the same problem, but `request_redraw` is called from within `RedrawEventsCleared`. Trying to conditionally spin the event loop with `ControlFlow::Wait`.

**Comment by parasyte** [comment]
Also this looks similar to #987

**Comment by alvinhochun** [comment]
> 
> 
> I copied your snippet into crate (with little modifications) and it waits for events.
> https://github.com/zakarumych/winit-wait-bug
> 
> What additional info I can provide to you?

I tried this code with `winit` master, but it doesn't reproduce the issue. I can see in the output that `RedrawRequested` and `redrawing!` does appear after every `MainEventsCleared`. Do you need to do anything special to trigger the issue?

**Comment by zakarumych** [comment]
> RedrawRequested and redrawing! does appear after every MainEventsCleared

But new sets of those appear only when I move a mouse, press a key or some other even occurs.
Just tried same code on Arch and it's still reproducible (originally I tried it on Windows).

**Comment by alvinhochun** [comment]
> But new sets of those [...] appear only when I move a mouse, press a key or some other even occurs

Is this not the expected behaviour? For each sets of events, the event handler is called with events in this order:

1. NewEvents(start_cause)
2. DeviceEvent { .. }, ..
3. UserEvent(_), ..
4. WindowEvent { .. }, ..
5. MainEventsCleared
6. RedrawRequested(_), ..
7. RedrawEventsCleared

Since you call `request_redraw` in `MainEventsCleared` (step 5), the redraw is queued and then immediately emitted afterwards (step 6). `ControlFlow::Wait` will then cause `winit` to wait for new events before going through the about steps again.

It sounds like you're expecting `request_redraw` will trigger a new sets of events and that the `RedrawRequested` event will cross over the next `NewEvents`. I don't see why this should be the case. Is it mentioned anywhere in the docs that I might have missed?

**Comment by zakarumych** [comment]
Yes, I guess it can be an expected behavior if `request_redraw` is called in `MainEventsCleared`.
But it is the same if `request_redraw` is called in `RedrawRequested` or `RedrawEventsCleared`, i.e. after last redraw_request is triggered and new one is requested but it still waits for something else to happen.

I don't know if this was mentioned anywhere in the docs, maybe I should do something else to achieve desired behavior - similar to invalidating window after redrawing it on Windows or calling `window.requestAnimationFrame(callback)` in the `callback` on web platform.

**Comment by filnet** [comment]
The `request_redraw` documentation states that [This function can cause `RedrawRequested` events to be emitted after `Event::MainEventsCleared` but before `Event::NewEvents` ...].It is not meant to carry over to the next event processing iteration.








    Le dimanche 27 septembre 2020 à 12:04:16 UTC+2, Zakarum <notifications@github.com> a écrit :  
 
 


Yes, I guess it can be an expected behavior if request_redraw is called in MainEventsCleared.
But it is the same if request_redraw is called in RedrawRequested or RedrawEventsCleared, i.e. after last redraw_request is triggered and new one is requested but it still waits for something else to happen.

I don't know if this was mentioned anywhere in the docs, maybe I should do something else to achieve desired behavior - similar to invalidating window after redrawing it on Windows or calling window.requestAnimationFrame(callback) in the callback on web platform.

**Comment by alvinhochun** [comment]
> 
> 
> Yes, I guess it can be an expected behavior if `request_redraw` is called in `MainEventsCleared`.
> But it is the same if `request_redraw` is called in `RedrawRequested` or `RedrawEventsCleared`, i.e. after last redraw_request is triggered and new one is requested but it still waits for something else to happen.

At least we can now agree that the current behaviour of `request_redraw` in `MainEventsCleared` is expected. There are two reasons I specifically referred to `MainEventsCleared` in my comments:

1. It is what the original report and the example codes focused on.
2. I do not have a clear answer to how `winit` should behave if `request_redraw` is called in `RedrawRequested` or `RedrawEventsCleared`. I know for sure that the web backend does not properly handle `request_redraw` being called anywhere after `MainEventsCleared` and before `NewEvents` but I do not know what the correct behaviour should be.

@kchibisov Do you have any input on this issue?

**Comment by kchibisov** [comment]
>@kchibisov Do you have any input on this issue?

Not really, other than winit should become more consistent here when it should send new redraw requested.

I can also say that when event loop starts processing events on Wayland, if you ask for redraw events cleared, it'll arrive on the 'next' even loop processing tick, which should differ from all platforms in winit, I guess.


In general, if you want to consult how redraw expected should be used there's an extensive discussion wrt its design https://github.com/rust-windowing/winit/issues/1041 . If the current docs are not clear when to call and what to expect, they should be updated, since the issue I've linked mentions `redraw_requested`.

**Comment by alvinhochun** [comment]
> I can also say that when event loop starts processing events on Wayland, if you ask for redraw events cleared, it'll arrive on the 'next' even loop processing tick, which should differ from all platforms in winit, I guess.

Even on or before `MainEventsCleared`? Is it due to practical limitations on Wayland, or is it just not implemented in the "usual" way yet?

> 
> In general, if you want to consult how redraw expected should be used there's an extensive discussion wrt its design #1041 . If the current docs are not clear when to call and what to expect, they should be updated, since the issue I've linked mentions `redraw_requested`.

Thanks, I'll read that issue when I have the time.

**Comment by filnet** [comment]
We might be in [XY problem](https://en.wikipedia.org/wiki/XY_problem) situation :)

Anyways..

One way to trigger/force a new event loop iteration in `ControlFlow::Wait` mode is to switch to `ControlFlow::Poll` mode (instead of calling `request_redraw()`) and then, in the next iteration, switch back to `ControlFlow::Wait` if needed.

**Comment by kchibisov** [comment]
>Even on or before MainEventsCleared? Is it due to practical limitations on Wayland, or is it just not implemented in the "usual" way yet?

no, that's how it is implemented, could be done the other way around. But winit lacks precise docs on what should be done and when, and I hate hunting random issues to get answers. But in my understand of how asking for a frame works, is that you call is right before you do drawing yourself, so calling it not from redraw requested event doesn't make any sense, and so in that case it'll properly enqueue event into event loop and make you process it on the next loop tick.

**Comment by zakarumych** [comment]
> One way to trigger/force a new event loop iteration in ControlFlow::Wait mode is to switch to ControlFlow::Poll mode (instead of calling request_redraw()) and then, in the next iteration, switch back to ControlFlow::Wait if needed.

Wouldn't `ControlFlow::Poll` cause multiple empty event iterations (`NewEvents`, `MainEventsCleared`, `RedrawEventsCleared`)  without `RedrawRequested` events on some platforms?

**Comment by QuantumBadger** [comment]
I've just been caught out by this, and found this ticket after spending a while debugging. The consensus above seems to be that this behaviour is expected, however to me personally it seems quite surprising that a pending `GlutinEvent::RedrawRequested` event doesn't wake up the event loop, and may only fire e.g. when the user next moves the mouse.

The behaviour also seems to differ on Windows and Linux -- on Linux it wakes up, on Windows it doesn't.

In any case, the workaround from @filnet works and was easy enough to implement: https://github.com/QuantumBadger/Speedy2D/commit/d0b881278aab3dbec68d6f199cdd79d89a3d61e0

**Comment by aevyrie** [comment]
I ran into this today, also expecting `request_redraw()` to trigger another winit loop when using `Wait`.

Another workaround is to create a custom event, and send a custom event instead of using the redraw method.
